Models/MuZero_agent_2.py

```diff
@@ -584,88 +589,136 @@ def player_to_play(player_num, action):
+        roots_cpp.prepare(config.ROOT_EXPLORATION_FRACTION, noises, value_prefix_pool, policy_logits_pool)
         
-        action = [int(self.select_action(root[i])) for i in range(self.NUM_ALIVE)]
-        
-        # Masking only if training is based on the actions taken in the environment.
-        # Training in MuZero is mostly based on the predicted actions rather than the real ones.
-        # network_output["policy_logits"], action = self.maskInput(network_output["policy_logits"], action)
+        # Output for root node
+        hidden_state_pool = network_output["hidden_state"] 
+
+        self.run_batch_mcts(roots_cpp, hidden_state_pool)  # set up nodes to be able to find and select actions 
+        roots_distributions = roots_cpp.get_distributions()
+        actions = [] 
+        # This variable controls if distributions is randomly created, such as during the very first loop, however it doesnt look like it always being True impacts anything 
+        start_training = True 
+        temp = self.visit_softmax_temperature() # controls the way actions are chosen
+        for i in range(self.NUM_ALIVE):
+            deterministic = False # False = sample distribution, True = argmax 
+            if start_training:
+                distributions = roots_distributions[i]
+            else:
+                #random distributions if training has just started 
+                distributions = np.ones(config.ACTION_DIM)
+            action, entropy = self.select_action(distributions,temperature=temp,deterministic=deterministic) 
+            actions.append(action)
 
         # Notes on possibilities for other dimensions at the bottom
         self.num_actions += 1
        
-        return action, network_output["policy_logits"]
-
-    def batch_expand_node(self, node: Node, to_play: int, network_output):
-        node.to_play = to_play
-        node.hidden_state = network_output["hidden_state"][to_play]
-        node.reward = network_output["reward"][to_play]
+        return actions, network_output["policy_logits"]
+    
+    @staticmethod
+    def select_action(visit_counts, temperature=1, deterministic=True):
+        """select action from the root visit counts.
+        Parameters
+        ----------
+        temperature: float
+            the temperature for the distribution
+        deterministic: bool
+            True -> select the argmax
+            False -> sample from the distribution
+        """
+        action_probs = [visit_count_i ** (1 / temperature) for visit_count_i in visit_counts]
+        total_count = sum(action_probs)
+        action_probs = [x / total_count for x in action_probs]
+        if deterministic:
+            action_pos = np.argmax([v for v in visit_counts])
+        else:
+            action_pos = np.random.choice(len(visit_counts), p=action_probs)
 
-        policy_probs = network_output["policy_logits"][to_play].numpy()
+        count_entropy = entropy(action_probs, base=2)
```

Comment:
Again. lets calculate this if we are going to use it. If not, no need.

Reply:
see above